add hw05-aop

aspanfilov opened the pull request.

aspanfilov committed:
add hw05-aop create project folder

aspanfilov committed:
hw05

aspanfilov committed:
add hw05 - aop

aspanfilov committed:
hw05-atm
Итоговый вариант

kataus commented on hw05-aop/src/main/java/ru/aop/TestLoggingCreator.java:
```diff
@@ -0,0 +1,59 @@
+
+            if (this.loggedMethods.contains(method)) {
+                System.out.println(getMethodLog(method, args));
+            } else if (isAnnotationPresent(method, Log.class)) {
```
Хорошее решение с ленивой загрузкой списка методов. Основная проблема - если метод не помечен как логируемый - мы постоянно будем перепроверять. Если есть желание оставить ленивый анализ - то лучше сделать обертку, которая будет по методу (уже проверенному) говорить - логируемый или нет.

aspanfilov replied:
Принято

kataus commented on hw05-aop/src/main/java/ru/aop/TestLoggingCreator.java:
```diff
@@ -0,0 +1,59 @@
+        }
+
+        private boolean isAnnotationPresent(Method method, Class<? extends Annotation> annotationClazz) throws NoSuchMethodException {
+            Method methodImpl = this.testLoggingClazz.getMethod(method.getName(), method.getParameterTypes());
```
А зачем? methodImpl чем-то отличается от method?

aspanfilov replied:
мethod - это же метод из интерфейса - он без аннотации. А тут мы достаем именно метод из конкретного класса - имплементации интерфейса. Пометка @Log указана в методе класса реализации, а не в методе интерфейса. Поэтому так.

kataus replied:
принято

aspanfilov committed:
edit hw05.
Исправление ведение кэширования для ранее считанных методов и аннотаций логирования к ним.
